drivers/flash/spi_nor.c

```diff
@@ -540,6 +546,104 @@ static int spi_nor_wrsr(const struct device *dev,
 	return ret;
 }
 
+#if DT_HAS_COMPAT_STATUS_OKAY(mxicy_mx25r_spi)
+
+/**
+ * @brief Read the configuration register.
+ *
+ * @note The device must be externally acquired before invoking this
+ * function.
+ *
+ * @param dev Device struct
+ *
+ * @return the non-negative value of the status register, or an error code.
+ */
+#define MX25R_CMD_RDCR 0x15
+static int mx25r_rdcr(const struct device *dev)
+{
+	uint16_t reg;
+	int ret = spi_nor_cmd_read(dev, MX25R_CMD_RDCR, &reg, sizeof(reg));
+
+	if (ret == 0) {
+		ret = reg;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Write the configuration register.
+ *
+ * @note The device must be externally acquired before invoking this
+ * function.
+ *
+ * @param dev Device struct
+ * @param cr  The new value of the configuration register
+ *
+ * @return 0 on success or a negative error code.
+ */
+static int mx25r_wrcr(const struct device *dev,
+			uint16_t cr)
+{
+	/* The configuration register bytes on the Macronix MX25R devices are
+	 * written using the Write Status Register command where the configuration
+	 * register bytes are written as two extra bytes after the status register.
+	 * First read out the current status register to preserve the value.
+	 */
+	int sr = spi_nor_rdsr(dev);
+
+	if (sr < 0) {
+		return sr;
+	}
+
+	int ret = spi_nor_cmd_write(dev, SPI_NOR_CMD_WREN);
+
+	if (ret == 0) {
+		uint8_t data[] = {
+			sr,
+			cr & 0xFF,	/* Configuration register 1 */
+			cr >> 8		/* Configuration register 2 */
+		};
+
+		ret = spi_nor_access(dev, SPI_NOR_CMD_WRSR, NOR_ACCESS_WRITE, 0, data,
+			sizeof(data));
+		spi_nor_wait_until_ready(dev);
+	}
+
+	return ret;
+}
+
+static int mx25r_configure(const struct device *dev)
+{
+	/* Low-power/high perf mode is second bit in configuration register 2 */
+	enum { MX25R_LH_SWITCH_BIT = 9 };
+
+	acquire_device(dev);
+
+	/* Read current configuration register */
+	int current_cr = mx25r_rdcr(dev);
+
+	if (current_cr < 0) {
+		return current_cr;
+	}
+
+	const bool use_high_perf = DT_PROP(DT_DRV_INST(0), mxicy_high_perf_mode);
+
+	LOG_DBG("Use high performance mode? %d", use_high_perf);
+	WRITE_BIT(current_cr, MX25R_LH_SWITCH_BIT, use_high_perf);
+	int ret = mx25r_wrcr(dev, current_cr);
```

Comment:
You could check if the configuration register already has the required value and avoid the write in such case.

Reply:
Sure, but is it worth the extra code to avoid one spi command? I'm not that good with bit-flipping and the macros we have available, but wouldn't we need something like this?

```cpp
if (current_cr & BIT(MX25R_LH_SWITCH_BIT) != use_high_perf << MX25R_LH_SWITCH_BIT) {
        ...
}
```

This might be shorter, but I'm not sure it's easier to understand:

```cpp
if (current_cr >> MX25R_LH_SWITCH_BIT & 0x1 != use_high_perf) {
        ...
}
```

Reply:
I thought about something like:
```
	required_cr = (current_cr & ~BIT(MX25R_LH_SWITCH_BIT)
		    | (use_high_perf ? BIT(MX25R_LH_SWITCH_BIT) : 0);
	if (required_cr != current_cr) {
		int ret = mx25r_wrcr(dev, current_cr);
```
I think it is worth it. We'll have few more CPU instructions instead of two SPI transactions (reading of the status register and writing of the status and configuration registers) that are unneeded for the default ultra low power mode (i.e. usually).

Reply:
Thanks, and with the `WRITE_BIT` macro for updating the `required_cr` it should be readable as well :)